flink-streaming-java/src/main/java/org/apache/flink/streaming/api/graph/StreamGraphTranslator.java

```diff
@@ -0,0 +1,93 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.api.graph;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.dag.Transformation;
+
+import java.util.Collection;
+
+/**
+ * A {@code Translator} is responsible for translating a given {@link Transformation}
+ * to its runtime implementation depending on the execution mode it is being executed.
+ * Currently, the execution mode can be either
+ * {@link org.apache.flink.streaming.api.RuntimeExecutionMode#STREAMING STREAMING} or
+ * {@link org.apache.flink.streaming.api.RuntimeExecutionMode#BATCH BATCH}.
+ *
+ * @param <OUT> The type of the output elements of the transformation being translated.
+ * @param <T> The type of transformation being translated.
+ */
+@Internal
+public interface StreamGraphTranslator<OUT, T extends Transformation<OUT>> {
+
+	/**
+	 * Translates a given {@link Transformation} to its runtime implementation for
+	 * {@link org.apache.flink.streaming.api.RuntimeExecutionMode#BATCH BATCH-style} execution.
+	 *
+	 * @param transformation The transformation to be translated.
+	 * @param context The translation context.
+	 * @return The ids of the nodes corresponding to this transformation in the transformation graph.
+	 */
+	Collection<Integer> translateForBatch(
+			final T transformation,
+			final Context context);
+
+	/**
+	 * Translates a given {@link Transformation} to its runtime implementation for
+	 * {@link org.apache.flink.streaming.api.RuntimeExecutionMode#STREAMING STREAMING-style} execution.
+	 *
+	 * @param transformation The transformation to be translated.
+	 * @param context The translation context.
+	 * @return The ids of the nodes corresponding to this transformation in the transformation graph.
+	 */
+	Collection<Integer> translateForStreaming(
+			final T transformation,
+			final Context context);
+
+	/**
+	 * A context giving the necessary information for the translation of a given transformation.
+	 */
+	interface Context {
+
+		/**
+		 * Returns the {@link StreamGraph} being created as the transformations
+		 * of a pipeline are translated to their runtime implementations.
+		 */
+		StreamGraph getStreamGraph();
+
+		/**
+		 * Returns the ids of the nodes in the {@link StreamGraph} corresponding to the
+		 * provided transformation.
+		 *
+		 * @param transformation the transformation whose nodes' ids we want.
+		 * @return The requested ids.
+		 */
+		Collection<Integer> getTransformationIds(final Transformation<?> transformation);
```

Comment:
Should maybe be `getStreamNodeIds()`, because it doesn't return transformation its.